tests/testcore/functional_test_base.go

```diff
@@ -244,12 +245,21 @@ func (s *FunctionalTestBase) SetupSuiteWithCluster(clusterConfigFile string, opt
 func (s *FunctionalTestBase) SetupTest() {
 	s.checkTestShard()
 	s.initAssertions()
+	s.testTimeout = newTestTimeout(s.T(), 2*time.Minute)
 }
 
 func (s *FunctionalTestBase) SetupSubTest() {
 	s.initAssertions()
 }
 
+func (s *FunctionalTestBase) TeardownTest() {
+	s.testTimeout.cancel()
+}
+
+func (s *FunctionalTestBase) SetTestTimeout(d time.Duration) {
+	s.testTimeout.set(d)
+}
```

Comment:
A test that needs more time can extend the timeout by calling this function.

(side effect: we can easily see which tests are very slow)